### DISCUSSION TITLE
How to configure Modern Fortran for Fortran 77 development on a legacy real-time system

### DISCUSSION THREAD

**Discussion by magol** [krvajal/vscode-fortran-support#1138] [General]
Hi,

I’m working on a legacy real-time system that uses Fortran IV and 77 code, and I’m facing some challenges in configuring the Modern Fortran extension to work within the constraints of this environment. Here’s a summary of the setup:

- Fortran IV/77 only: The system uses Fortran IV/77 exclusively, and it is not compatible with any newer versions of Fortran.
- Compiler via Telnet: The Fortran compiler is accessible only via Telnet, and there is no local compilation possible.
- Files via FTP: The code files are accessed via FTP using a non-standard path syntax. As an alternative, we also have an API, developed in-house, that can be used to manage file transfers.
- Debugging via Telnet: The debugging tool is primitive and depends on information output by the compiler to determine memory addresses for functions and variables. This tool is used to step through the code and set breakpoints.
- Pre-processor: The code contains many constants in the form `<FOO>`, which are replaced by a number or string during pre-processing.

I understand that achieving full integration for all of these requirements may not be feasible, but I’m looking for guidance on how to configure the extension to support my workflow as much as practically possible. Specifically, I’m looking for advice on:

- Ensuring the extension recognizes and supports Fortran IV/77 code exclusively.
- Configuring the extension to work with a remote compiler over Telnet.
- Integrating file access through FTP (or possibly our custom API).
- Setting up debugging with the Telnet-based debugger, which relies on compiler output for memory addresses.
- Handling the pre-processing of constants in the form `<FOO>` within the extension or providing a workaround for this.

Any suggestions or workarounds to streamline my workflow would be greatly appreciated.

**Comment by gnikit** [comment]
> * Ensuring the extension recognizes and supports Fortran IV/77 code exclusively.

For F77 you can use the `FixedFormFortran` syntax highlighting and that should do the trick. Obviously custom syntax outside the Standard or older versions of Fortran will be a hit or miss depending on your codebase.


> * Configuring the extension to work with a remote compiler over Telnet.

This I do not know. A google search pointed me to this SO [post](https://stackoverflow.com/questions/71802649/similar-to-ibmi-as400-can-we-do-telnet-on-visual-studio-code). Maybe it is of value to you.

> * Integrating file access through FTP (or possibly our custom API).

Not sure about this. Probably the SO post I mentioned above would be applicable here as well. If you have SSH in your remote system you can use the SSH extension on the remote machine and work this way. I am guessing that anything outside of that is not really supported by VS code.


> * Setting up debugging with the Telnet-based debugger, which relies on compiler output for memory addresses.

That is probably not possible. VS code needs a debug adaptor to populate the debugger's GUI. You would have to write the adaptor yourself and of course we could ship it with the extension so that you can use it, but again that would only work only if the code runs **on** vscode either locally or through the remote extensions on the remote.

> * Handling the pre-processing of constants in the form `<FOO>` within the extension or providing a workaround for this.

With the exception of `fypp` we don't support custom pre-processors because over the years there have been so many and how they mutate sources is not necessarily something that can be mapped to the original sources and serve accurate diagnostics during linting.

-----

From the sounds of it you are working on a very customised environment/codebase and you won't be able to use any open-source solution out of the box. It might be worth sharing more details about the remote system, compiler used, debugger, build system and also the system paths which you mentioned are somehow custom.

At this point my advice would be to develop locally, make the code as F77 compliant as possible. Try and get it to compile with gfortran using the legacy options and then send the sources over FTP/SSH to the remove machine and iterate this way.

If you can't share certain information publicly you can get my email from the commit history of the repo and carry this conversation in private. However, my feeling is that there are no easy solutions.
